NormalMapNode: Add basic support for normal unpacking.

castano opened the pull request:
Normal maps are often compressed using the BC5 and EAC_RG formats. This is currently not supported by three.js as the shaders expect the sampled normal map values to contain RGB components representing the packed XYZ coordinate.

To address this issue I propose we add a `unpackNormalMode` attribute to the NormalMapNode object. When this attribute is true, the generated shader code computes the Z component of the normal by projecting the XY components over the hemisphere.

The `unpackNormalMode` attribute is set automatically when the texture is using the RG or BC5 formats. A future PR will add support for EAC formats and will also set the attribute appropiately for EAC_RG formats. For other use cases, such as using `unpackNormalMode` for ASTC and UASTC textures, we would need to introduce another mechanism to signal how textures are encoded.

*This contribution is funded by [Ludicon](https://ludicon.com)*

castano committed:
Add NormalMapNode projectZ attribute.

When the projectZ attribute is true this computes the Z component of the normal by projecting the XY components over the hemisphere.

Mugen87 commented on src/nodes/display/NormalMapNode.js:
```diff
@@ -57,14 +58,28 @@ class NormalMapNode extends TempNode {
 
+		if ( projectZ ) {
+
+			normalMap = vec3( normalMap.xy, sqrt( saturate( float( 1.0 ).sub( dot( normalMap.xy, normalMap.xy ) ) ) ) );
```
Do you mind moving this code into`src/nodes/utils/Packing.js` into a new function? It might be handy elsewhere.

```js
/**
 * Unpacks RG-packed normals by reconstructing their Z component by projecting it to the hemisphere.
 *
 * @tsl
 * @function
 * @param {Node<vec3>} node - The color to unpack.
 * @return {Node<vec3>} The direction.
 */
export const unpackRGNormal = ( normal ) => vec3( normal.xy, sqrt( saturate( float( 1.0 ).sub( dot( normal.xy, normal.xy ) ) ) ) );
```
We can then use it in `NormalMapNode` like so:
```js
if ( useRGPacking === true ) {

    normalMap = unpackRGNormal( normalMap );

}

Mugen87 commented on src/nodes/accessors/MaterialNode.js:
```diff
@@ -235,6 +235,12 @@ class MaterialNode extends Node {
 				node = normalMap( this.getTexture( 'normal' ), this.getCache( 'normalScale', 'vec2' ) );
 				node.normalMapType = material.normalMapType;
 
+				if ( material.normalMap.projectZ ) {
```
When thinking about this: Could you also check for specific RG texture formats and only then set the node property to `true`? In this way we need no additional texture property.

castano replied:
I went with the userData approach. Let me know if that looks good to you.

Mugen87 replied:
I wonder if we should introduce `Texture.packing` and remove the `Normal` term from the constants. We could then have `NoPacking, `RGPacking` and `GAPacking`. 

I could image that next to normals other texture types could also pack their data. The question is if we can implement this a bit deeper in the node system. Maybe in `TextureNode`.

It would be interesting to hear the opinion of others about this PR. Would you be okay if we label this PR as a "Draft PR" for now? I have the feeling we need to discuss the API a bit more in the team/community before making a decision.

Mugen87 replied:
> I'm not sure enabling the Z reconstruction is the right thing to do for all RG texture formats.

My idea was to only do this for specific map types. So for a normal map, a format like `bc5-rg-snorm` must hold packed RG data, otherwise it would unusable for a normal map. This is something that could be implemented in the node classes so loaders don't have to worry about this detail.

castano replied:
Yeah, now that I think about it, it could make sense to do it based on the format. This happens inside the NormalMapNode, so I think it's safe to assume the texture being sampled is indeed a normal map.

The explicit unpack mode allows supporting ASTC LA normal maps or DXT5nm and in these cases it's not possible to infer the packing mode from the format alone.

I don't see an option to convert this into a Draft PR, but I'd be happy to hear other people's feedback.

Mugen87 replied:
> The explicit unpack mode allows supporting ASTC LA normal maps or DXT5nm and in these cases it's not possible to infer the packing mode from the format alone.

@donmccurdy I'll ping you directly since you have done so much work in context of texture compression.

Similar to `Texture.colorSpace`, what do you think about introducing a new property `Texture.packing` that describes the packing strategy of a texture. The default would be `NoPacking` (as a pendant to `NoColorSpace`). 

Packing strategies usually apply on certain texture types so probably most of the new constants for `Texture.packing`  would be quite specific (like `RGNormalPacking` or `GANormalPacking`). But that is a similar situation like with `Texture.colorSpace` where certain constants only apply to color textures.

So...what do you think about this approach? Do you have maybe alternatives in mind?

donmccurdy replied:
I wouldn't be opposed to adding a `Texture#packing` property, I think @bhouston had proposed something along those lines once? A few other thoughts:

1. If we see `texture.format = THREE.RGFormat` assigned to a normal map slot, possibly that is enough in this case?

2. In the newer TSL/NodeMaterial APIs with WebGPURenderer, is a property on THREE.Texture still accessible? Or could the Texture instance have been wrapped in other nodes, obscuring the underlying Texture? If so, we could handle this by a new node type, or a material property similar to `material.normalMapType`, instead.

castano replied:
OK, I've been looking at the possibility of enabling the RGNormalPacking mode automatically based on the texture format. It could be something along these lines:

```
if ( material.normalMap.format == THREE.RGFormat || material.normalMap.format == RED_GREEN_RGTC2_Format ) {
	node.unpackNormal = NormalRGPacking;
}
```

Let me know if it's worth adding this. The default behavior could still be overridden with the user data attribute: `normalMap.userData.unpackNormal`.

Note, this only handles RG textures and BC5. I think there's no identifier for EAC_RG11 in three.js. If support for it is added in the future, this could be handled here as well.

We could also add support for signed RG textures and eliminate the `2*x-1` unpacking in the normal map node, but barely anybody uses the signed formats, so I think it's not worth it.

I'm ambivalent about adding a `Texture#packing` property. I can experiment with that if you think that's the way to go.

donmccurdy replied:
I suspect checking `.format` would make sense for WebGLRenderer but not for TSL in WebGPURenderer. Since the focus of this PR is WebGPURenderer, we can skip that for now, but I wouldn't be opposed to adding this support in WebGLRenderer later.

castano replied:
This should work with either renderer. The main caveats are that it requires the format attribute to be set (which may not be the case if it's an external texture), and it does not support EAC_RG, since that format is not natively supported in three.js.

Happy to include it for now if you think it makes sense to pave the way in that direction.

castano commented on src/nodes/display/NormalMapNode.js:
```diff
@@ -57,14 +58,28 @@ class NormalMapNode extends TempNode {
+		 * @type {bool}
+		 * @default false
+		 */
+		this.projectZ = false;
```
Thanks for the feedback. Back in the day it was common to pack normal maps using the `G` and `A` components using the DXT5 format. This is again common with ASTC and it's luminance-alpha encoding. Perhaps this should be named `usePacking` and set to `''` or `'rg'`, so that it can be extended in the future.

I can add these to `constants.js`:

```
/**
 * No normal map packing.
 * 
 * @type {string}
 * @constant
 */
export const NoNormalPacking = '';

/**
 * Normal RG packing.
 * 
 * @type {string}
 * @constant
 */
export const NormalRGPacking = 'rg';
```

castano replied:
I went ahead and added support for both RG packing and LA packing.

castano committed:
Add unpack normal helpers to Packing.js

castano committed:
Add normal packing mode constants to constants.js

castano committed:
Import normal packing mode constants.

castano committed:
Rename projectZ to unpackNormal.

Change type from bool to string and use packing mode constants.
Print errors when trying to use incompatible packed normals and normal modes.

castano committed:
Material sets unpackNormal attribute based on normal map user data attribute.

castano committed:
Fix lint errors.

castano committed:
Fix linter error.

castano commented on src/nodes/display/NormalMapNode.js:
```diff
@@ -57,14 +58,48 @@ class NormalMapNode extends TempNode {
-		const { normalMapType, scaleNode } = this;
+		const { normalMapType, scaleNode, unpackNormal } = this;
 
 		let normalMap = this.node.mul( 2.0 ).sub( 1.0 );
```
This could use the `colorToDirection` function in Packing.js

That said, in the case of normal maps it may make more sense to use:

```
let normalMap = this.node.mul( 2.0 ).sub( 254.0 / 255.0 );
```

As that prevents lighting discontinuities on flat normals along uv seams where the tangents do not match exactly.

kristiker commented on src/nodes/utils/Packing.js:
```diff
@@ -19,3 +20,25 @@ export const directionToColor = ( node ) => nodeObject( node ).mul( 0.5 ).add( 0
+ * @param {Node<vec3>} normal - The packed normal to unpack.
+ * @return {Node<vec3>} The unpacked normal.
+ */
+export const unpackRGNormal = ( normal ) => vec3( normal.xy, sqrt( saturate( float( 1.0 ).sub( dot( normal.xy, normal.xy ) ) ) ) );
```
Imo this should be called reconstructZ and it should be channel ordering unaware. I.e. you should pass normal x,y yourself.

castano replied:
How about this:

```
export const reconstructZ = ( xy ) => vec3( xy, sqrt( saturate( float( 1.0 ).sub( dot( xy, xy ) ) ) ) );
```

Since this is generating a syntax tree, I'd rather pass a vec2 than two separate x,y values, as that will result in cleaner shader code generation.

castano committed:
Replace unpackNormal functions with reconstructZ.

castano committed:
Merge branch 'dev' into normalmap-projectz

castano commented:
I've put together a demo that uses the changes I'm proposing here. You can check it out here: https://ludicon.com/sparkjs/gltf-demo/

castano committed:
Apply suggested renaming.

castano committed:
Rename unpackNormalZ and update invocation.

castano committed:
Or better, keep unpackNormal function and rename attribute.

Mugen87 commented:
@castano Sorry for the delay but do you think you can share one normal map that uses BC5 or EAC_RG compression? I think we could add a new mesh using such a texture in [webgpu_materials](https://threejs.org/examples/webgpu_materials) so we have at least one test case for this code path.

castano commented:
@Mugen87 Certainly! Sorry it took me so long. I could not find any KTX encoder that supported EAC_RG out of the box, so I had to write my own.

Here's a normal map from: https://ambientcg.com/view?id=Ground080 in BC5 and EAC_RG format:
[Ground080_1K_NormalGL.zip](https://github.com/user-attachments/files/23246730/Ground080_1K_NormalGL.zip)

Let me know if this works, or whether your would prefer a different texture or a different container format.

Mugen87 commented on src/nodes/accessors/MaterialNode.js:
```diff
@@ -235,6 +235,12 @@ class MaterialNode extends Node {
 
+				if ( material.normalMap.userData && material.normalMap.userData.unpackNormalMode ) {
+
+					node.unpackNormalMode = material.normalMap.userData.unpackNormalMode;
```
I have reconsidered this bit. Do you mind changing this check by explicitly testing for compatible texture formats? In this way, we don't have to store data in `Texture.userData` which means if loaders or apps properly configure compressed normals maps they will automatically work with `WebGPURenderer`.

castano replied:
OK. Done!

castano committed:
Merge remote-tracking branch 'upstream/dev' into normalmap-projectz

castano committed:
Automatically apply NormalRGPacking mode to RG and BC5 texture nodes used as normal maps.

castano committed:
Remove the use of normalMap.userData.unpackNormalMode to override the node's unpackNormalMode.
